File: DemoObjC/RootViewController.m
//
//  RootViewController.m
//  ToasterDemoObjC
//
//  Created by Antoine Cœur on 2019/4/4.
//  Copyright © 2019 Suyeol Jeon. All rights reserved.
//

#import "RootViewController.h"
@import Toaster;

@interface RootViewController ()

@end

@interface RespondingButton : UIButton<UIKeyInput>

@property(assign, nonatomic) BOOL hasText;
@property(assign, nonatomic) UITextAutocorrectionType autocorrectionType;

@end

@implementation RootViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
    [button setTitle:@"Show" forState:UIControlStateNormal];
    [button sizeToFit];
    [button addTarget:self action:@selector(showButtonTouchUpInside) forControlEvents:UIControlEventTouchUpInside];
    button.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin;
    button.center = CGPointMake(self.view.center.x, 75);
    [self.view addSubview:button];
    
    RespondingButton *keyboardButton = [RespondingButton buttonWithType:UIButtonTypeSystem];
    [keyboardButton setTitle:@"Toggle keyboard" forState:UIControlStateNormal];
    [keyboardButton sizeToFit];
    [keyboardButton addTarget:self action:@selector(keyboardButtonTouchUpInside:) forControlEvents:UIControlEventTouchUpInside];
    keyboardButton.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin;
    keyboardButton.center = CGPointMake(self.view.center.x, 125);
    [self.view addSubview:keyboardButton];
    
    UIButton *shadowButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [shadowButton setTitle:@"Toggle shadow" forState:UIControlStateNormal];
    [shadowButton sizeToFit];
    [shadowButton addTarget:self action:@selector(shadowButtonTouchUpInside:) forControlEvents:UIControlEventTouchUpInside];
    shadowButton.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin;
    shadowButton.center = CGPointMake(self.view.center.x, 155);
    [self.view addSubview:shadowButton];
    
    [self configureAppearance];
    [self configureAccessibility];
}

- (void)configureAppearance {
    ToastView *appearance = [ToastView appearance];
    appearance.backgroundColor = UIColor.lightGrayColor;
    appearance.textColor = UIColor.blackColor;
    appearance.font = [UIFont boldSystemFontOfSize: 16];
    appearance.textInsets = UIEdgeInsetsMake(15, 20, 15, 20);
    appearance.bottomOffsetPortrait = 100;
    appearance.cornerRadius = 20;
    appearance.maxWidthRatio = 0.7;
}

- (void)configureAccessibility {
    [ToastCenter default].isSupportAccessibility = YES;
}

- (void)showButtonTouchUpInside {
    [[[Toast alloc] initWithText:@"Basic Toast" delay:0 duration:Delay.Short] show];
    [[[Toast alloc] initWithAttributedText:[[NSAttributedString alloc] initWithString:@"AttributedString Toast" attributes:@{NSBackgroundColorAttributeName: [UIColor yellowColor]}] delay:0 duration:Delay.Short] show];
    [[[Toast alloc] initWithText:@"You can set duration. `Delay.Short` means 2 seconds.\n`Delay.Long` means 3.5 seconds." delay:0 duration:Delay.Long] show];
    [[[Toast alloc] initWithText:@"With delay, Toaster will be shown after delay." delay:1 duration:5] show];
}

- (void)keyboardButtonTouchUpInside:(RespondingButton *)sender {
    if (sender.isFirstResponder) {
        [sender resignFirstResponder];
    } else {
        [sender becomeFirstResponder];
    }
}

- (void)shadowButtonTouchUpInside:(UIButton *)sender {
    ToastView *appearance = [ToastView appearance];
    if (appearance.shadowColor) {
        appearance.shadowColor = nil;
        appearance.shadowOpacity = 0;
        appearance.shadowOffset = CGSizeZero;
        appearance.shadowRadius = 0;
    } else {
        appearance.shadowColor = [UIColor blackColor];
        appearance.shadowOpacity = 0.2;
        appearance.shadowOffset = CGSizeMake(0, 1);
        appearance.shadowRadius = 10;
    }
}

@end

@implementation RespondingButton

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    self.hasText = YES;
    self.autocorrectionType = UITextAutocorrectionTypeNo;
    return self;
}

- (BOOL)canBecomeFirstResponder {
    return YES;
}

- (void)insertText:(nonnull NSString *)text {}
- (void)deleteBackward {}

@end

File: Demo/RootViewController.swift
import UIKit
import Toaster
import Photos

final class RootViewController: UIViewController {

  override func viewDidLoad() {
    super.viewDidLoad()
    
    let button = UIButton(type: .system)
    button.setTitle("Show", for: .normal)
    button.sizeToFit()
    button.addTarget(self, action: #selector(self.showButtonTouchUpInside), for: .touchUpInside)
    button.autoresizingMask = [.flexibleBottomMargin, .flexibleLeftMargin, .flexibleRightMargin]
    button.center = CGPoint(x: view.center.x, y: 75)
    self.view.addSubview(button)

    let keyboardButton = RespondingButton(type: .system)
    keyboardButton.setTitle("Toggle keyboard", for: .normal)
    keyboardButton.sizeToFit()
    keyboardButton.addTarget(self, action: #selector(self.keyboardButtonTouchUpInside), for: .touchUpInside)
    keyboardButton.autoresizingMask = [.flexibleBottomMargin, .flexibleLeftMargin, .flexibleRightMargin]
    keyboardButton.center = CGPoint(x: view.center.x, y: 125)
    self.view.addSubview(keyboardButton)

    let imagePickerButton = UIButton(type: .system)
    imagePickerButton.setTitle("Open imagePicker", for: .normal)
    imagePickerButton.sizeToFit()
    imagePickerButton.addTarget(self, action: #selector(self.imagePickerButtonTouchUpInside), for: .touchUpInside)
    imagePickerButton.autoresizingMask = [.flexibleBottomMargin, .flexibleLeftMargin, .flexibleRightMargin]
    imagePickerButton.center = CGPoint(x: view.center.x, y: 155)
    self.view.addSubview(imagePickerButton)
    
    let shadowButton = UIButton(type: .system)
    shadowButton.setTitle("Toggle shadow", for: .normal)
    shadowButton.sizeToFit()
    shadowButton.addTarget(self, action: #selector(self.shadowButtonTouchUpInside), for: .touchUpInside)
    shadowButton.autoresizingMask = [.flexibleBottomMargin, .flexibleLeftMargin, .flexibleRightMargin]
    shadowButton.center = CGPoint(x: view.center.x, y: 185)
    self.view.addSubview(shadowButton)

    self.configureAppearance()
    self.configureAccessibility()
  }

  func configureAppearance() {
    let appearance = ToastView.appearance()
    appearance.backgroundColor = .lightGray
    appearance.textColor = .black
    appearance.font = .boldSystemFont(ofSize: 16)
    appearance.textInsets = UIEdgeInsets(top: 15, left: 20, bottom: 15, right: 20)
    appearance.bottomOffsetPortrait = 100
    appearance.cornerRadius = 20
    appearance.maxWidthRatio = 0.7
  }

  func configureAccessibility() {
    ToastCenter.default.isSupportAccessibility = true
  }

  @objc dynamic func showButtonTouchUpInside() {
    Toast(text: "Basic Toast").show()
    Toast(attributedText: NSAttributedString(string: "AttributedString Toast", attributes: [NSAttributedString.Key.backgroundColor: UIColor.yellow])).show()
    Toast(text: "You can set duration. `Delay.short` means 2 seconds.\n" +
      "`Delay.long` means 3.5 seconds.",
          duration: Delay.long).show()
    Toast(text: "With delay, Toaster will be shown after delay.", delay: 1, duration: 5).show()
  }

  @objc dynamic func keyboardButtonTouchUpInside(sender: RespondingButton) {
    if sender.isFirstResponder {
      sender.resignFirstResponder()
    } else {
      sender.becomeFirstResponder()
    }
  }

  @objc dynamic func imagePickerButtonTouchUpInside(sender: UIButton) {
    pickImage()
  }
  
  @objc dynamic func shadowButtonTouchUpInside(sender: UIButton) {
    let appearance = ToastView.appearance()
    if appearance.shadowColor == nil {
      appearance.shadowColor = .black
      appearance.shadowOpacity = 0.2
      appearance.shadowOffset = CGSize(width: 0, height: 1)
      appearance.shadowRadius = 10
    } else {
      appearance.shadowColor = nil
      appearance.shadowOpacity = 0
      appearance.shadowOffset = .zero
      appearance.shadowRadius = 0
    }
  }

  private func pickImage() {
    let authorizationStatus = PHPhotoLibrary.authorizationStatus()
    if authorizationStatus == .denied {
      print("You didn't grant photo album permission. Please check your settings.")
      return
    } else if authorizationStatus == .notDetermined {
      // prompt the user about permissions before opening photo album
      if #available(iOS 11.0, *) {
        // nothing to do: no permission request needed on iOS 11+
      } else {
        PHPhotoLibrary.requestAuthorization { success in
          if success == .authorized {
            DispatchQueue.main.async {
              self.pickImage()
            }
          }
        }
        return
      }
    }
    let mediaPicker = UIImagePickerController()
    mediaPicker.delegate = self
    present(mediaPicker, animated: true, completion: nil)
  }
}

extension RootViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
  func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
    Toast(text: "Hello, Great image.").show()
    picker.dismiss(animated: true, completion: nil)
  }

  func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
    picker.dismiss(animated: true) {
      print("User cancelled")
    }
  }
}

class RespondingButton: UIButton, UIKeyInput {
  override var canBecomeFirstResponder: Bool {
    return true
  }
  var hasText: Bool = true
  var autocorrectionType: UITextAutocorrectionType = .no
  func insertText(_ text: String) {}
  func deleteBackward() {}
}
